test/sslapitest.c

```diff
@@ -9229,6 +9229,93 @@ static int test_pluggable_group(int idx)
 
     return testresult;
 }
+
+/* This function triggers encode, decode and sign functions
+ * of the artificial "hmacsig" algorithm implemented in tls-provider
+ */
+static int create_cert_key(int idx, char *certfilename, char *privkeyfilename)
+{
+    EVP_PKEY_CTX * evpctx = EVP_PKEY_CTX_new_from_name(libctx,
+                             (idx==0)?"xorhmacsig":"xorhmacsha2sig", NULL);
+    EVP_PKEY *pkey = NULL;
+    X509 * x509 = X509_new();
+    X509_NAME * name = NULL;
+    FILE *f = NULL;
+    int ret = 1;
+
+    if (!TEST_ptr(evpctx) ||
+        !TEST_true(EVP_PKEY_keygen_init(evpctx)) ||
+        !TEST_true(EVP_PKEY_generate(evpctx, &pkey)) ||
+        !TEST_ptr(pkey) ||
+        !TEST_ptr(x509) ||
+        !TEST_true(ASN1_INTEGER_set(X509_get_serialNumber(x509), 1)) ||
+        !TEST_true(X509_gmtime_adj(X509_get_notBefore(x509), 0)) ||
+        !TEST_true(X509_gmtime_adj(X509_get_notAfter(x509), 31536000L)) ||
+        !TEST_true(X509_set_pubkey(x509, pkey)) ||
+        !TEST_ptr(name = X509_get_subject_name(x509)) ||
+        !TEST_true(X509_NAME_add_entry_by_txt(name, "C",  MBSTRING_ASC,
+                           (unsigned char *)"CH", -1, -1, 0)) ||
+        !TEST_true(X509_NAME_add_entry_by_txt(name, "O",  MBSTRING_ASC,
+                           (unsigned char *)"test.org", -1, -1, 0)) ||
+        !TEST_true(X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC,
+                           (unsigned char *)"localhost", -1, -1, 0)) ||
+        !TEST_true(X509_set_issuer_name(x509, name)) ||
+        !TEST_true(X509_sign(x509, pkey, EVP_sha1())) ||
+        !TEST_ptr(f = fopen(privkeyfilename, "wb")) ||
+        !TEST_true(PEM_write_PrivateKey(f, pkey, NULL, NULL, 0, NULL, NULL)) ||
+        fclose(f) ||
+        !TEST_ptr(f = fopen(certfilename, "wb")) ||
+        !TEST_true(PEM_write_X509(f, x509)) ||
+        fclose(f))
+        ret = 0;
+
+    if (pkey) EVP_PKEY_free(pkey);
+    if (x509) X509_free(x509);
```

Comment:
You also need to free `evpctx`.  That omission should be the reason for at least one of the leaks detected in https://github.com/openssl/openssl/actions/runs/3815985801/jobs/6491331364

Reply:
(also, you don't have to test if the pointers are NULL before freeing them.  All our freeing functions check for NULL and simply do nothing in that case, just like `free()`...  or if you find one that doesn't, you've found a bug that should be corrected)

Reply:
so changed in https://github.com/openssl/openssl/pull/19312/commits/f716ba4cebb9c0586b746e99554b6f92aada9501